# NumPy 

NumPy provides efficient storage, it is faster and nevertheless, provides efficient way to handling and processing datas. Now importing numpy,

In [1]:
import numpy as np

Creating an array.

In [2]:
arr = np.array([[3,2,6,7]], np.int8)
arr 

array([[3, 2, 6, 7]], dtype=int8)

By using np.int__ , here we can put the integer bit. By which we can easily handle memory management. we can put 32bit or 64bit for wokring with huge numbers.

In [3]:
arr.shape 

(1, 4)

If we need to know informations about the 'arr', then we can use arr.shape. Using this it will show us number of rows and columns.

In [4]:
arr.dtype

dtype('int8')

' arr.dtype ' shows us which data type it is. To know more,
https://numpy.org/doc/stable/reference/

Now, assigning values in array,

In [5]:
arr[0, 2] = 69
arr

array([[ 3,  2, 69,  7]], dtype=int8)

There are 5 ways we can create a numpy array.
# 1) Array Creation using python structures.

In [6]:
listarray = np.array([[1,2,4],[2,4,6],[7,8,9]])
listarray

array([[1, 2, 4],
       [2, 4, 6],
       [7, 8, 9]])

In [7]:
listarray.dtype

dtype('int32')

In [8]:
listarray.shape

(3, 3)

In [9]:
listarray.size

9

# 2) Creating array using intrinsic numpy array 
means, zeros, ones, arange etc

In [10]:
# Using zeros
zeros = np.zeros((2,5))
zeros

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

Using (2,5) means, it will create 2 rows and 5 columns. And all the elements will be zero.

In [11]:
# Using arange
rng = np.arange(15)
rng

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

It will create a numpy array from 0 to 15.

In [12]:
lspace = np.linspace(1,5,12)
lspace

array([1.        , 1.36363636, 1.72727273, 2.09090909, 2.45454545,
       2.81818182, 3.18181818, 3.54545455, 3.90909091, 4.27272727,
       4.63636364, 5.        ])

It will give us 12 elements which will be equally linearly spaced from 1 to 5

In [13]:
emp = np.empty((4,6))
emp

array([[6.23042070e-307, 4.67296746e-307, 1.69121096e-306,
        8.90095411e-307, 1.89146896e-307, 7.56571288e-307],
       [4.00544531e-307, 1.11261095e-306, 3.56043054e-307,
        1.37961641e-306, 2.22518251e-306, 1.33511969e-306],
       [6.23037317e-307, 9.34609790e-307, 8.45593934e-307,
        9.34600963e-307, 1.86921143e-306, 6.23061763e-307],
       [2.22522053e-306, 1.11261095e-306, 1.24610383e-306,
        1.33511290e-306, 8.34445562e-308, 3.91792476e-317]])

It will return 4 rows and each row will have 6 columns. All the elements will be random.

In [14]:
emp_like = np.empty_like(lspace)
emp_like

array([1.        , 1.36363636, 1.72727273, 2.09090909, 2.45454545,
       2.81818182, 3.18181818, 3.54545455, 3.90909091, 4.27272727,
       4.63636364, 5.        ])

In [15]:
ide = np.identity(45)
ide

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

It will return identity matrix.

In [19]:
array = np.arange(99)
array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98])

In [22]:
array = array.reshape(3,33)
array

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
        32],
       [33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48,
        49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64,
        65],
       [66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81,
        82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97,
        98]])

Making it a 1D array

In [24]:
array = array.ravel()
array

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98])

# Axis in NumPy

1D array has number of axis: 1 [Axis 0]
2D aray has number of axis: 2  [Axis 0, Axis 1]
3D aray has number of axis: 3  [Axis 0, Axis 1, Axis 2]

In [26]:
x = [[1,2,3],[4,5,6],[7,1,0]]
ar = np.array(x)
ar

array([[1, 2, 3],
       [4, 5, 6],
       [7, 1, 0]])

In [27]:
ar.sum(axis = 0)

array([12,  8,  9])

axis = 0, means it will go through the elements 1 * 1, 2 * 1, 3 * 1. 
axis = 1, means it will go through the elements 1 * 1, 1 * 2, 1 * 3.

In [28]:
ar.sum(axis = 1)

array([ 6, 15,  8])

Transpose of a matrix

In [29]:
ar.T

array([[1, 4, 7],
       [2, 5, 1],
       [3, 6, 0]])

To iterate through the matrix, use flat() function.

In [31]:
for item in ar.flat:
    print(item)

1
2
3
4
5
6
7
1
0


We can also know the dimension of a matrix using 'ndim'

In [32]:
ar.ndim

2

In [33]:
ar.size

9

We can find out how much space our matrix consuming by using 'nbytes'

In [34]:
ar.nbytes

36

In [40]:
# Creating one dimensional array
oneD = np.array([1,2,44,32,121,18,20])

'arg' methods for One Dimensional array

In [37]:
# Finding the index of maximum element
oneD.argmax()

4

In [38]:
# Finding the element of minimum element
oneD.argmin()

0

In [39]:
# Sorted array but gives us index
oneD.argsort()

array([0, 1, 5, 6, 3, 2, 4], dtype=int64)

In [41]:
ar

array([[1, 2, 3],
       [4, 5, 6],
       [7, 1, 0]])

In [42]:
ar.argmin()

8

In [43]:
ar.argmax()

6

In [44]:
ar.argmax(axis=0)

array([2, 1, 1], dtype=int64)

Here, max element is 7 in (1,4,7) that's why 2. In (2,5,1), max element is 5 that's why 1.

In [45]:
ar.argmax(axis = 1)

array([2, 2, 0], dtype=int64)

In [47]:
ar.argsort(axis = 0)

array([[0, 2, 2],
       [1, 0, 0],
       [2, 1, 1]], dtype=int64)

# Matrix operations 

In [50]:
ar2 = np.array([[1,2,1],
                [4,0,6],
                [8,1,0]])

# Adding two matrices
ar+ar2

array([[ 2,  4,  4],
       [ 8,  5, 12],
       [15,  2,  0]])

In [51]:
# Multiplying 
ar*ar2

array([[ 1,  4,  3],
       [16,  0, 36],
       [56,  1,  0]])

In [52]:
# Subtraction
ar-ar2

array([[ 0,  0,  2],
       [ 0,  5,  0],
       [-1,  0,  0]])

In [54]:
np.sqrt(ar)

array([[1.        , 1.41421356, 1.73205081],
       [2.        , 2.23606798, 2.44948974],
       [2.64575131, 1.        , 0.        ]])

In [55]:
ar.sum()

29

In [56]:
ar.max()

7

In [57]:
ar.min()

0

Finding numbers

In [58]:
np.where(ar>5)

(array([1, 2], dtype=int64), array([2, 0], dtype=int64))

In [60]:
type(np.where(ar>5))

tuple

In [61]:
# Counts the non-zero elements
np.count_nonzero(ar)

8

In [62]:
np.nonzero(ar)

(array([0, 0, 0, 1, 1, 1, 2, 2], dtype=int64),
 array([0, 1, 2, 0, 1, 2, 0, 1], dtype=int64))

In [64]:
ar[1,2] = 0
np.nonzero(ar)

(array([0, 0, 0, 1, 1, 2, 2], dtype=int64),
 array([0, 1, 2, 0, 1, 0, 1], dtype=int64))

# Proof of memory efficiency of NumPy

In [67]:
import sys

In [68]:
#Python array
py_ar = [23,11,69,33]

#NumPy array
np_ar =np.array(py_ar)

In [69]:
sys.getsizeof(1) * len(py_ar)

112

In [71]:
np_ar.itemsize * np_ar.size

16

Hence, it is seen that python array(py_ar) taking 112 bytes where NumPy array(np_ar) taking only 16 bytes.

https://docs.scipy.org/doc//numpy-1.10.4/reference/arrays.ndarray.html#array-attributes